### PR TITLE
RLP-1087: iterate ChannelTestCase

### PR DESCRIPTION
**Pull request by mortelli**
This PR partly implements issue [RLP-1087](https://jirainfuy.atlassian.net/browse/RLP-1087), which is about iterating and improving on the `ChannelTestCase` class.

The second part of this task (named parameters in the constructor function) was not implemented due to complexity VS readability issues, plus the already significant size of this PR (more details in the issue).

### Overview
1. `ChannelTestCase` class was renamed to `ChannelState`, and now represents an expected channel state to be used in tests for verification purposes.
2. Directories were restructured in the following manner: utilities functions remain in `utils`, common classes and constants reside in `common`, and tests are now split into `channels/` and `payments/` (`bootstrapping` doesn't have a home).
3. Channel tests were split into `open`, `close` and `deposit`. 
    1. individual `params` interfaces were created in each of these to encapsulate SDK parameters.
4. Some missing verifications were added (i.e. checking channels are as expected from the other side).

### CODE REVIEW

File: test/cases/bootstrapping.test.ts [hunk 1]
```diff
@@ -1,46 +1,33 @@
...
+        'should be able to boot up an advanced node setup',
+        async () => {
+            env = await setupTestEnvironment(advanced as SetupJson);
+            await verifyEnv(env);
+        },
+        SETUP_TIMEOUT
+    );
 });
-
-async function verifyEnv(env: LuminoTestEnvironment): Promise<void> {
```
Comment (mortelli) [inline test/cases/bootstrapping.test.ts:38]: moved to `utils`

File: test/cases/channel/close.test.ts [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+    token: string;
+    partner: string;
+}
+
+describe('channel close', () => {
+    let nodes: Dictionary<LuminoNode>;
+    let env: LuminoTestEnvironment;
+
+    beforeAll(async () => {
+        env = await setupTestEnvironment(mediated);
```
Comment (mortelli) [inline test/cases/channel/close.test.ts:20]: `close` now uses the `mediated` topology so that we have 2 channels that can be closed, and one close operation executed from each side of a channel

File: test/cases/bootstrapping.test.ts [hunk 3]
```diff
@@ -1,46 +1,33 @@
...
+import advanced from '../../topologies/advanced.json';
+import { SetupJson } from '../../src/types/setup';
+import setupTestEnvironment from '../../src';
+import { LuminoTestEnvironment } from '../../src/types/lumino-test-environment';
+import { verifyEnv } from '../utils';
+import { SETUP_TIMEOUT, TEARDOWN_TIMEOUT } from '../common';
 
-const SETUP_TIMEOUT = 5 * 60 * 1000; // 5 minutes
-const TEARDOWN_TIMEOUT = 1 * 60 * 1000; // 1 minute
+describe('bootstrapping', () => {
```
Comment (mortelli) [inline test/cases/bootstrapping.test.ts:9]: this test is still a bit homeless. it will probably be reworked or outright removed at some point, as discussed with @JONAF2103 and @mauricioirace

File: test/cases/channel/close.test.ts [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+    afterAll(async () => {
+        await env.stop();
+    }, TEARDOWN_TIMEOUT);
+
+    it(
+        'from initiator node',
+        async () => {
+            // deposit for node "initiator"
+            const initiatorDeposit = toWei(
+                mediated.channels[0].participant1.deposit // should be inferred
```
Comment (mortelli) [inline test/cases/channel/close.test.ts:33]: these should eventually be fixed under [RLP-1084](https://jirainfuy.atlassian.net/browse/RLP-1084)

File: test/cases/channel/deposit.test.ts [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+    let env: LuminoTestEnvironment;
+
+    beforeAll(async () => {
+        env = await setupTestEnvironment(p2p);
+        nodes = env.nodes as Dictionary<LuminoNode>;
+    }, SETUP_TIMEOUT);
+
+    afterAll(async () => {
+        await env.stop();
+    }, TEARDOWN_TIMEOUT);
```
Comment (mortelli) [inline test/cases/channel/deposit.test.ts]: i've seen a lot of repeated code across tests with these statements. it can probably be improved upon, or at least extracted

File: test/cases/channel/open.test.ts [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+        await env.stop();
+    }, TEARDOWN_TIMEOUT);
+
+    test(
+        'initiator node, 0 tokens',
+        async () => {
+            const params: openParams = {
+                token: tokenAddresses.LUM,
+                amount: toWei(0),
+                partner: '0x8645315E490A05FeE7EDcF671B096E82D9b616a4', // too arbitrary, unrelated to topology
```
Comment (mortelli) [inline test/cases/channel/open.test.ts]: through [RLP-1088](https://jirainfuy.atlassian.net/browse/RLP-1088) we should eventually have open test cases for existing nodes in the topology, as well as valid addresses like these, which do not correspond to running topology nodes (while we're at it—let's test for invalid addresses too)

File: test/cases/payments/mediated.test.ts [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+    const initiatorDeposit = toWei(mediated.channels[0].participant1.deposit);
+    const mediatorInitiatorDeposit = toWei(
+        mediated.channels[0].participant2.deposit
+    );
+    const mediatorTargetDeposit = toWei(
+        mediated.channels[1].participant1.deposit
+    );
+    const targetDeposit = toWei(mediated.channels[1].participant2.deposit);
+
+    const paymentAmount = toWei(1);
```
Comment (mortelli) [inline test/cases/payments/mediated.test.ts:30]: somewhat ugly, but it's needed to calculate the expected final balances on each side of both channels, for all 3 nodes. 

it could be moved to the actual `test` statement, but i think it muddies the code a little bit, which is already long enough.

open to suggestions here

File: test/cases/payments/p2p.test.ts [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+    let nodes: Dictionary<LuminoNode>;
+    let env: LuminoTestEnvironment;
+
+    // starting deposit for nodes
+    const initiatorDeposit = toWei(
+        p2p.channels[0].participant1.deposit // should be inferred
+    );
+    const targetDeposit = toWei(
+        p2p.channels[0].participant2.deposit // should be inferred
+    );
```
Comment (mortelli) [inline test/cases/payments/p2p.test.ts:26]: these actually need to be outside the `test` statements, since separate payments (back and forth) occur across the 2

File: test/cases/payments/p2p.test.ts [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+                amount: initiatorPaymentAmount,
+            };
+
+            await nodes.initiator.client.sdk.makePayment({
+                tokenAddress: params.token,
+                partnerAddress: params.partner,
+                amountOnWei: params.amount,
+            });
+
+            await sleep(5000); // should not be necessary
```
Comment (mortelli) [inline test/cases/payments/p2p.test.ts:56]: all of thse go under [RLP-1086](https://jirainfuy.atlassian.net/browse/RLP-1086)

File: test/common/index.ts [hunk 10]
```diff
@@ -0,0 +1,31 @@
+export const SETUP_TIMEOUT = 5 * 60 * 1000; // 5 minutes
+export const TEARDOWN_TIMEOUT = 1 * 60 * 1000; // 1 minute
+export const TEST_TIMEOUT = 1 * 60 * 1000; // 1 minute
```
Comment (mortelli) [inline test/common/index.ts]: my python brain cells say "these should be static class attributes"

File: test/cases/payments/p2p.test.ts [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+import { tokenAddresses, toWei } from '../../../src/util/token';
+import {
+    SETUP_TIMEOUT,
+    TEARDOWN_TIMEOUT,
+    ChannelState,
+    State,
+    TEST_TIMEOUT,
+} from '../../common';
+
+interface paymentParams {
```
Comment (mortelli) [inline test/cases/payments/p2p.test.ts:10]: this one is repeated from `mediated`, but i can't decide were to place it honestly. maybe a common file inside the payments tc dir?

File: test/cases/channel/close.test.ts [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+    TEARDOWN_TIMEOUT,
+    ChannelState,
+    State,
+    TEST_TIMEOUT,
+} from '../../common';
+
+interface closeParams {
+    token: string;
+    partner: string;
+}
```
Comment (mauricioirace) [inline test/cases/channel/close.test.ts:13]: Maybe we could change ths for:
```suggestion
interface closeParams {
    tokenAddress: string;
    partnerAddress: string;
}
```
and add that interface to lumino-js-sdk.d.ts so we can call closeChannel(params) more easily

File: test/cases/channel/deposit.test.ts [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+    ChannelState,
+    State,
+    TEST_TIMEOUT,
+} from '../../common';
+
+interface depositParams {
+    token: string;
+    amount: number;
+    partner: string;
+}
```
Comment (mauricioirace) [inline test/cases/channel/deposit.test.ts:14]: Same as above

File: test/cases/channel/close.test.ts (hunk 4 above)
Comment (mauricioirace) [inline test/cases/channel/close.test.ts:33]: I disagree with this approach. We are sharing (due to performance limitations) same containers for all the tests in a test suite. If some other test changes the balance of the channel, the initial balance will no longer be  `mediated.channels[0].participant1.deposit`. 

I'd replace this with a call to the getChannel endpoint, to prevent issues in the future

File: test/cases/channel/close.test.ts [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+                params.partner,
+                expected
+            );
+
+            // repeat verification from "mediator" node
+
+            // deposit for node "mediator"
+            const mediatorDeposit = toWei(
+                mediated.channels[0].participant2.deposit // should be inferred
+            );
```
Comment (mauricioirace) [inline test/cases/channel/close.test.ts:65]: same as above

File: test/utils/index.ts [hunk 15]
```diff
@@ -1,52 +1,40 @@
...
-    this.state = channelState;
-  }
+export async function verifyEnv(env: LuminoTestEnvironment): Promise<void> {
+    expect(Object.keys(env.nodes).length).toBe(2);
+    const firstNode = env.nodes.node0 as LuminoNode;
+    const secondNode = env.nodes.node1 as LuminoNode;
+    const firstAddr = (await firstNode.client.sdk.getAddress()).our_address;
+    const secondAddr = (await secondNode.client.sdk.getAddress()).our_address;
+    expect(isEmpty(firstAddr)).toBe(false);
+    expect(isEmpty(secondAddr)).toBe(false);
```
Comment (mauricioirace) [inline test/utils/index.ts:21]: What if the env have more than two nodes? or less?

**Review by mauricioirace — COMMENTED** [review]
Just some comments

File: test/cases/channel/close.test.ts (hunk 12 above)
Comment (mortelli) [reply test/cases/channel/close.test.ts:13]: > Maybe we could change ths for:
> 
> and add that interface to lumino-js-sdk.d.ts so we can call closeChannel(params) more easily

i like this idea. it could both simplify the SDK calls as well as make the tests cleaner.

if @JONAF2103 agrees, i'll create an issue. we'll have to think about how best to design these "params" classes or interfaces so that they make sense across all the SDK operations.

File: test/cases/channel/deposit.test.ts (hunk 13 above)
Comment (mortelli) [reply test/cases/channel/deposit.test.ts:14]: reply [here](https://github.com/mauricioirace/lumino-tests/pull/6#discussion_r630473553)

File: test/utils/index.ts (hunk 15 above)
Comment (mortelli) [reply test/utils/index.ts:21]: indeed, this poses a problem. not just the quantity of the nodes, but the names as well. 

_if_ we want to keep testing the setting up of the environments—and i think we should—we need code that can adapt to any number of nodes, and with arbitrary names.

this means we might need a bit more structure than just JSON files, or maybe we need to be rethink the way we are parsing them.

this `verifyEnv` function came as a result of extracting duplicated code in the bootstrapping test. as you know, as a team we're still not fully convinced that this test should be kept.

i didn't want to delete the test, so this is the result, which is not ideal. i was hoping we could tackle these issues through [RLP-1084](https://jirainfuy.atlassian.net/browse/RLP-1084).

File: test/cases/channel/close.test.ts (hunk 4 above)
Comment (mortelli) [reply test/cases/channel/close.test.ts:33]: > I disagree with this approach. We are sharing (due to performance limitations) same containers for all the tests in a test suite. If some other test changes the balance of the channel, the initial balance will no longer be `mediated.channels[0].participant1.deposit`.
> 
> I'd replace this with a call to the getChannel endpoint, to prevent issues in the future

let me start off by saying that i am not happy with this line (and its cousins) for several reasons. it's hard-coded rather than inferred, plus it uses a numeric index. so for sure we will need a better alternative.

having said that, what do you mean by "if some other test changes the balance"? 

if you mean the deposit amount written in the JSON, then i see where you're going with this. but if you mean another test file, then it shouldn't affect this one. afaik each test has its separate set-up and teardown. do we agree on this?

whatever the case may be, i'd like to touch upon the idea of calling an endpoint to get this information. i'd much rather opt for having a robust topology component that eliminates this need, chiefly because:
1. API calls are expensive and will slow down tests.
2. we should be able to trust our own topologies (be they encoded in JSON files or something else entirely). i don't think we need to take a defensive approach against ourselves as long as the code is usable and decently adaptable.

on a more philosophical note, and purely as an opinion: from a testing perspective we should always aim to compare expected results against actual functionality. this means that we should have expected data and then compare it to—in this case—an API call result. 

if we need to call the API to obtain data (like "channel deposit") which will be, in turn, compared against _more_ API results, it is now more likely the correctness of the test will be compromised, since it is using more of the elements that are being tested. 

i understand that in some cases there's just no getting around it, but what i'm getting at here is that the less we depend on API calls for tests, the more robust they'll be.

@JONAF2103 thoughts?

File: test/cases/channel/close.test.ts (hunk 14 above)
Comment (mortelli) [reply test/cases/channel/close.test.ts:65]: see reply [here](https://github.com/mauricioirace/lumino-tests/pull/6#discussion_r630524178)

File: test/cases/channel/close.test.ts (hunk 4 above)
Comment (JONAF2103) [reply test/cases/channel/close.test.ts:33]: Well i have some comments about this:

1. About the numeric index it's because the defined topology, actually an alternative to that doesn't exists since in the simple topology we define general properties and a total number of nodes. Since those nodes are not named you must to call them by index, that's why. The responsible for the tests is who should check if the topology is simple or advanced, so i disagree in that point. We should anyway to create a validator or something to check if those indexes are accessible or not before running any test.
2. I'm not aware about the "performance limitations" that @mauricioirace is talking about but we shouldn't be storing any state in the sdk or either any of the wrapper objects, actually if we want to check the balance we should always call the node with the sdk and check the actual balance of the channel in that moment, those calls are not heavy at all, actually the channel call is really lightweight.

File: test/cases/channel/deposit.test.ts (hunk 5 above)
Comment (JONAF2103) [reply test/cases/channel/deposit.test.ts]: I agree but let's not change all this into a black box code please :) we should see the limit between improve the test code and write a jest test, the hooks code probably will be pretty similar on a lot of tests but we need to keep in mind that probably for some test we will need to make changes.

File: test/cases/channel/open.test.ts (hunk 6 above)
Comment (JONAF2103) [reply test/cases/channel/open.test.ts]: Agree

File: test/cases/payments/mediated.test.ts (hunk 7 above)
Comment (JONAF2103) [reply test/cases/payments/mediated.test.ts:30]: I think is ok but we should start moving all these common values to constants on something called `commons` or something like that to use a reference of it instead of being calling web3 each time and calculating the amount to use. This applies to all the other cases where we use 0, 1 or some common value.

File: test/common/index.ts (hunk 10 above)
Comment (JONAF2103) [reply test/common/index.ts]: Actually here we should use objects instead, constants is the best way but to have a sense of structure we should use objects to create categories of constants and use them, for example:

```typescript
export const timeouts = {
     setup: 5 * 60 * 1000, // 5 minutes
     teardown: 1 * 60 * 1000, // 1 minute
     test: 1 * 60 * 1000 // 1 minute
};
```
and then you can access those values like:
```typescript
import {timeouts} from '...';

timeouts.setup
...
timeouts.teardown
...
timeouts.test
```
Same with all the other constants.

File: test/utils/index.ts (hunk 15 above)
Comment (JONAF2103) [reply test/utils/index.ts:21]: I think when we create the validator for the topologies this will be solved and we will remove this function.

**Review by JONAF2103 — APPROVED** [review]
Looks good in general i just leave you some comments.

File: test/cases/payments/mediated.test.ts (hunk 7 above)
Comment (mauricioirace) [reply test/cases/payments/mediated.test.ts:30]: I agree we both of you. I like @JONAF2103 's idea.

File: test/cases/payments/p2p.test.ts (hunk 11 above)
Comment (mauricioirace) [reply test/cases/payments/p2p.test.ts:10]: Maybe we could add a test/types dir, or something like that. I'm not very sure either

File: test/cases/channel/close.test.ts (hunk 12 above)
Comment (mortelli) [reply test/cases/channel/close.test.ts:13]: created issue [RLP-1090](https://jirainfuy.atlassian.net/browse/RLP-1090) for this

File: test/cases/channel/close.test.ts (hunk 4 above)
Comment (mortelli) [reply test/cases/channel/close.test.ts:39]: >     1. About the numeric index it's because the defined topology, actually an alternative to that doesn't exists since in the simple topology we define general properties and a total number of nodes. Since those nodes are not named you must to call them by index, that's why. The responsible for the tests is who should check if the topology is simple or advanced, so i disagree in that point. We should anyway to create a validator or something to check if those indexes are accessible or not before running any test.

i agree, specially on your last point. i am fine with leaving this as-is until we iterate on the topology (or an eventual validator for it).

>     2. I'm not aware about the "performance limitations" that @mauricioirace is talking about but we shouldn't be storing any state in the sdk or either any of the wrapper objects, actually if we want to check the balance we should always call the node with the sdk and check the actual balance of the channel in that moment, those calls are not heavy at all, actually the channel call is really lightweight.

if we have no other choice but to call the SDK for obtaining some data, then that's that.

but, leaving performance aside: if through the specified topology we already know what the supposed balance (or deposit, etc.) should be, then why would we call the SDK to obtain this information?

i think that specially in tests, we should always trust our own test case data over what the SDK might say, and this is what i tried to convey in my last message: hard-coded, expected values make the tests more robust in comparison to calculated ones, moreso if they are calculated by the component we are testing.

so in summary, i have this question: why should we use the SDK to obtain a value we can deduct from the test-case data, such as the topology?

Comment (mortelli) [reply test/cases/channel/close.test.ts:33]: >     1. About the numeric index it's because the defined topology, actually an alternative to that doesn't exists since in the simple topology we define general properties and a total number of nodes. Since those nodes are not named you must to call them by index, that's why. The responsible for the tests is who should check if the topology is simple or advanced, so i disagree in that point. We should anyway to create a validator or something to check if those indexes are accessible or not before running any test.

i agree, specially on your last point. i am fine with leaving this as-is until we iterate on the topology (or an eventual validator for it).

>     2. I'm not aware about the "performance limitations" that @mauricioirace is talking about but we shouldn't be storing any state in the sdk or either any of the wrapper objects, actually if we want to check the balance we should always call the node with the sdk and check the actual balance of the channel in that moment, those calls are not heavy at all, actually the channel call is really lightweight.

if we have no other choice but to call the SDK for obtaining some data, then that's that.

but, leaving performance aside: if through the specified topology we already know what the supposed balance (or deposit, etc.) should be, then why would we call the SDK to obtain this information?

i think that specially in tests, we should always trust our own test case data over what the SDK might say, and this is what i tried to convey in my last message: hard-coded, expected values make the tests more robust in comparison to calculated ones, moreso if they are calculated by the component we are testing.

so in summary, i have this question: why should we use the SDK to obtain a value we can deduct from the test-case data, such as the topology?

File: test/cases/channel/deposit.test.ts (hunk 5 above)
Comment (mortelli) [reply test/cases/channel/deposit.test.ts]: > I agree but let's not change all this into a black box code please :) we should see the limit between improve the test code and write a jest test, the hooks code probably will be pretty similar on a lot of tests but we need to keep in mind that probably for some test we will need to make changes.

sure. let's leave it as it is for now. i don't think extracting this would make it "black box code", but i see your point

File: test/cases/payments/mediated.test.ts (hunk 7 above)
Comment (mortelli) [reply test/cases/payments/mediated.test.ts:30]: hmm, not sure if that would be too much abstraction, but i'm open to the idea for now. 

i'm willing to review such an implementation if you guys want to include it in some other PR

File: test/cases/payments/p2p.test.ts (hunk 11 above)
Comment (mortelli) [reply test/cases/payments/p2p.test.ts:10]: yes, but things like `depositParams` are really only used in the deposit test, so i believe that one should stay inside the test and not in a types file. 

unless of course we add it to the SDK code, then it would probably be best to move it

File: test/utils/index.ts (hunk 15 above)
Comment (mortelli) [reply test/utils/index.ts:21]: agreed

File: test/common/index.ts (hunk 10 above)
Comment (mortelli) [reply test/common/index.ts]: done! thank you very much for the suggestion

File: test/cases/channel/close.test.ts (hunk 4 above)
Comment (mauricioirace) [reply test/cases/channel/close.test.ts:33]: >i think that specially in tests, we should always trust our own test case data over what the SDK might say, and this is what i tried to convey in my last message: hard-coded, expected values make the tests more robust in comparison to calculated ones, moreso if they are calculated by the component we are testing.

I agree with this. Maybe we could replace beforeAll and afterAll by beforeEach and AfterEach, in order to use this approach. 
I don't dislike it. but note it will last a long time to run the tests this way.

Anyway remembering they are integration tests, maybe it isn't something so weird.

File: test/cases/payments/p2p.test.ts (hunk 11 above)
Comment (mauricioirace) [reply test/cases/payments/p2p.test.ts:10]: I'd then leave it this way, and update that when we tackle the corresponding issue

**Review by mauricioirace — APPROVED** [review]
